### ISSUE TITLE
Big loss using openchat tricks in transformers

### ISSUE THREAD

**Issue by SingL3** [imoneoi/openchat#74]
Hi, I am using openchat's tricks to do training on other models, like [internlm](https://huggingface.co/internlm/internlm-7b).
I didnt use this repo but base on the `transformers` lib.
I do the following modification:
1. Contructing the dataset as `GPT3/4 User` and `GPT3/4 Bot`.
2. Constructing the dataset which yield `input_ids` `labels` and `weights`
3. Modify the loss part
from 
```
loss = None
  if labels is not None:
  # Shift so that tokens < n predict n
  shift_logits = logits[..., :-1, :].contiguous()
  shift_labels = labels[..., 1:].contiguous()
  # Flatten the tokens
  loss_fct = CrossEntropyLoss()
  shift_logits = shift_logits.view(-1, self.config.vocab_size)
  shift_labels = shift_labels.view(-1)
  # Enable model parallelism
  shift_labels = shift_labels.to(shift_logits.device)
  loss = loss_fct(shift_logits, shift_labels)
```
to
```
        loss = None
        if labels is not None:
            # Shift so that tokens < n predict n
            shift_logits = logits[..., :-1, :].contiguous()
            shift_labels = labels[..., 1:].contiguous()
            # Flatten the tokens
            shift_logits = shift_logits.view(-1, self.config.vocab_size)
            shift_labels = shift_labels.view(-1)
            # Enable model parallelism
            shift_labels = shift_labels.to(shift_logits.device)
            
            shift_weights = weights[..., 1:].contiguous()
            shift_weights = shift_weights.view(-1)
            shift_weights = shift_weights.to(shift_logits.device)
            loss = weighted_cross_entropy(shift_logits, shift_labels, shift_weights)
```

Formerly, if i dont apply these modification, I will get a loss like `1.x` and now I get a loss like `100.x` .

Are there any other modification I should do?

**Comment by imoneoi** [comment]
@SingL3 the weighted cross-entropy sums up the losses of all tokens. We divide the final loss by the total number of tokens on all GPUs. You can check it here https://github.com/imoneoi/openchat/blob/master/ochat/training_deepspeed/train.py#L281

**Comment by SingL3** [comment]
I see. You compute the weighted average cross entropy.
Closing this. Thank you!

**Comment by jujeongho0** [comment]
Hi!
I have a question about your mention.

When I look at the code, it divides the final loss by the all_numseq, but isn't this the sum of the weighted losses, not the total number of tokens?

**Comment by SingL3** [comment]
@jujeongho0 
Check these lines:
https://github.com/imoneoi/openchat/blob/7df44963c9ec2cbb677a622db1de71cf5d4adbe4/ochat/data/generate_dataset.py#L55-L57

**Comment by jujeongho0** [comment]
Yes.
Doesn't that mean divides the final loss by the sum of the weighted loss, not the total number of tokens?

So, if I were to write that code as a base for ```transformers``` lib, following @SingL3's method, it might look like this.
```
import torch

SEQ_LEN = 100
VOCAB_SIZE = 40960
lm_logits = torch.randn(1, SEQ_LEN, VOCAB_SIZE)
labels = torch.randint(VOCAB_SIZE, (1, SEQ_LEN))
gpt3_weights = torch.zeros(1, SEQ_LEN)
gpt4_weights = torch.zeros(1, SEQ_LEN)

SOURCE_LEN = 50
labels[..., :SOURCE_LEN] = -100
gpt3_weights[..., SOURCE_LEN:] = 0.1
gpt4_weights[..., SOURCE_LEN:] = 1.0

shift_logits = lm_logits[..., :-1, :].contiguous()
shift_labels = labels[..., 1:].contiguous()
shift_gpt3_weights = gpt3_weights[..., 1:].contiguous()
shift_gpt4_weights = gpt4_weights[..., 1:].contiguous()

x = shift_logits.view(-1, shift_logits.size(-1))
y = shift_labels.view(-1)
gpt3_w = shift_gpt3_weights.view(-1)
gpt4_w = shift_gpt4_weights.view(-1)

gpt3_loss = (gpt3_w * torch.nn.functional.cross_entropy(x, y, reduction="none")).sum() # tensor(56.5003)
gpt4_loss = (gpt4_w * torch.nn.functional.cross_entropy(x, y, reduction="none")).sum() # tensor(565.0032)

weighted_gpt3_loss = gpt3_loss // sum(gpt3_w) # tensor(11.)
weighted_gpt4_loss = gpt4_loss // sum(gpt4_w) # tensor(11.)
```

Looking at the resulting values, doesn't it mean that the loss for the data in GPT4 and the data in GPT3 will eventually be the same, regardless of the weight?

If you see anything wrong with the above code, please let me know.
Thanks.

**Comment by SingL3** [comment]
@jujeongho0 
I am not sure whether I get it right in your code. Did you compute the loss for gpt3 and gpt4 separately?

In my implement, 

> 3. Modify the loss part
>    from
> 
> ```
> loss = None
>   if labels is not None:
>   # Shift so that tokens < n predict n
>   shift_logits = logits[..., :-1, :].contiguous()
>   shift_labels = labels[..., 1:].contiguous()
>   # Flatten the tokens
>   loss_fct = CrossEntropyLoss()
>   shift_logits = shift_logits.view(-1, self.config.vocab_size)
>   shift_labels = shift_labels.view(-1)
>   # Enable model parallelism
>   shift_labels = shift_labels.to(shift_logits.device)
>   loss = loss_fct(shift_logits, shift_labels)
> ```
> 
> to
> 
> ```
>         loss = None
>         if labels is not None:
>             # Shift so that tokens < n predict n
>             shift_logits = logits[..., :-1, :].contiguous()
>             shift_labels = labels[..., 1:].contiguous()
>             # Flatten the tokens
>             shift_logits = shift_logits.view(-1, self.config.vocab_size)
>             shift_labels = shift_labels.view(-1)
>             # Enable model parallelism
>             shift_labels = shift_labels.to(shift_logits.device)
>             
>             shift_weights = weights[..., 1:].contiguous()
>             shift_weights = shift_weights.view(-1)
>             shift_weights = shift_weights.to(shift_logits.device)
>             loss = weighted_cross_entropy(shift_logits, shift_labels, shift_weights)
> ```

It should be:
```
        loss = None
        if labels is not None:
            # Shift so that tokens < n predict n
            shift_logits = logits[..., :-1, :].contiguous()
            shift_labels = labels[..., 1:].contiguous()
            # Flatten the tokens
            # loss_fct = CrossEntropyLoss()
            shift_logits = shift_logits.view(-1, self.config.vocab_size)
            shift_labels = shift_labels.view(-1)
            # Enable model parallelism
            shift_labels = shift_labels.to(shift_logits.device)
            
            shift_weights = weights[..., 1:].contiguous()
            shift_weights = shift_weights.view(-1)
            shift_weights = shift_weights.to(shift_logits.device)
            loss = weighted_cross_entropy(shift_logits, shift_labels, shift_weights)
            loss = (1 / torch.sum(shift_weights)) * loss
```

Formerly, I didn't divide the sum of weights.
Check these lines:

https://github.com/imoneoi/openchat/blob/7df44963c9ec2cbb677a622db1de71cf5d4adbe4/ochat/training_deepspeed/train.py#L280-L282

**Comment by SingL3** [comment]
@jujeongho0 
Sry, I dont understand what is your point. The mean of loss, in my opinion, is somewhat like a `all_reduce` operation. And this should not influence the training.

**Comment by han508** [comment]
when labels is -100, its loss is 0, it can rule it out?
